### DISCUSSION TITLE
Timestamp Timezone Problem

### DISCUSSION THREAD

**Discussion by offizium-berndstorath** [laravel/framework#51161] [Q&A]
I get this in phpmyadmin

| time                | created_at          |
|---------------------|---------------------|
| 2024-04-22 00:40:11 | 2024-04-22 00:40:11 |

I get this using xdebug `$model->created_at->toString();`

| time                | created_at          |
|---------------------|---------------------|
| Mon Apr 22 2024 00:40:11 GMT+0200 | Mon Apr 22 2024 00:40:11 GMT+0200 |

I get this when calling the api

| time                | created_at          |
|---------------------|---------------------|
| 2024-04-22T02:40:09.000000Z | 2024-04-22T00:40:09.000000Z |

casts are 
```php
$casts = [
  'time' => 'datetime:Y-m-d\TH:i:s.u\Z',
  'created_at' => 'datetime:Y-m-d\TH:i:s.u\Z'
];
```

mysql and config timezones are +2 and Europe/Berlin respectively. Both are mysql data type `TIMESTAMP`

When setting the mysql timezone to utc this problem doesnt exist anymore. But this is not the fix i am hoping for

Why do the two times suddenly differ?

**Comment by ghost** [comment]
Set your timezone in laravel

    /*
    |--------------------------------------------------------------------------
    | Application Timezone
    |--------------------------------------------------------------------------
    |
    | Here you may specify the default timezone for your application, which
    | will be used by the PHP date and date-time functions. We have gone
    | ahead and set this to a sensible default for you out of the box.
    |
    */

    'timezone' => 'UTC',

**Reply by offizium-berndstorath** [reply]
Why tho? Both my server and database are in Europe/Berlin. And I already have a lot of data in the database and applications expecting the right timezone

**Reply by ghost** [reply]
Because that is how laravel works. You need to change your timezone in laravel config if the timezone is not UTC

**Reply by offizium-berndstorath** [reply]
I suppose I understood your answer wrong. I already set the timezone in the laravel config to Europe/Berlin

**Reply by ghost** [reply]
both time and create_at are timestamps in db? ... aaa yes you mentioned..

**Reply by ghost** [reply]
can you pls comment the casts and present what is the api response?

**Reply by offizium-berndstorath** [reply]
I already mentioned the casts in my post. the api response is
```json
{
    "created_at": "2024-04-22T00:40:09.000000Z",
    "time": "2024-04-22T02:40:09.000000Z"
}
```

**Reply by ghost** [reply]
I mean comment them out or delete/remove them to see the un-casted values.

**Reply by offizium-berndstorath** [reply]
```json
{
    "created_at": "2024-04-21T22:40:11.000000Z",
    "time": "2024-04-22 00:40:11"
}
```
This is a different database entry, thats why the times are entirely different

**Reply by ghost** [reply]
This seems like the db type is datetime not timestamp. And it gets converted to the timezone of the laravel app (+2 h)

**Reply by offizium-berndstorath** [reply]
No. Its timestamp. If i change it to datetime i get the same result

**Reply by ghost** [reply]
https://dev.mysql.com/doc/refman/8.0/en/datetime.html
![image](https://github.com/laravel/framework/assets/82870454/1fbf5723-e284-4f21-aa93-a37c932d08bd)

**Reply by offizium-berndstorath** [reply]
I know. Both show the same value in the database but differ when fetched through the api. this makes no sense. they are the same datatype so they should be the same every time

**Reply by ghost** [reply]
"The time zone can be set on a per-connection basis. As long as the time zone setting remains constant, you get back the same value you store. If you store a TIMESTAMP value, and then change the time zone and retrieve the value, the retrieved value is different from the value you stored"

**Reply by offizium-berndstorath** [reply]
I do not change the timezone. The problem is most definitely not the database

**Reply by ghost** [reply]
This is one of the reasons I prefer always UTC and datetime for storage and if needed convert it to local date time only on display.

**Reply by offizium-berndstorath** [reply]
Yes I understand. But this project already exists and is unfortunately to big to update. Is there a workaround?

**Reply by ghost** [reply]
I suggest you create a laravel project from scratch with ONLY ONE new table from scratch using a new db schema that will use timezone and timestamp columns.
If you get the same results then it is not the mysql.

Also how are these columns populated? automatic from mysql or from php?

**Reply by offizium-berndstorath** [reply]
Populated through laravel

**Reply by ghost** [reply]
Can you give the 
`show create table {your_table};`
query result in time and created_at regard?

**Reply by offizium-berndstorath** [reply]
```sql
CREATE TABLE `table` (
  `time` timestamp NOT NULL,
  `created_at` timestamp NULL DEFAULT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=13876 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
```

**Reply by ghost** [reply]
![image](https://github.com/laravel/framework/assets/82870454/7389d0b6-42e2-4cc0-9c2b-7e5bc24d9f21)

The last one is with timezone +3 in laravel. The first ones are with timezone UTC in laravel.

All the time I receive on dd the utc

array:2 [
  "time" => "2024-04-22T13:25:05.000000Z"
  "created_at" => "2024-04-22T13:25:05.000000Z"
]


and when I list

![image](https://github.com/laravel/framework/assets/82870454/f0363eef-953c-4e0d-b2fc-b7d228f3e56e)

So if I set timezone in laravel it saved on db +3h and when it retrieves it removes the 3 h from db.
My mysql has SYSTEM as timezone which is +3 GMT.

**Reply by offizium-berndstorath** [reply]
That means?

**Reply by ghost** [reply]
your time may be the updated at for some fields?

**Reply by offizium-berndstorath** [reply]
Why would it be? My full table schema already has a `updated_at` column

**Reply by ghost** [reply]
I just asked because you did not mention until now updated at. So next question then would be how are you populating the time column? Using timestamps = true in your model will auto-fill ONLY the updated_at and created_at which I see are ok on your side.

**Reply by offizium-berndstorath** [reply]
I am using `$table->timestamps();` in the migration.`time` is a from something different. but because of timings most of the time the same as `created_at`

**Reply by ghost** [reply]
I updated my first element
then the list became:
![image](https://github.com/laravel/framework/assets/82870454/d15a0e75-e5d4-4e2b-9114-b6b0541030a5)

So the first row was created with UTC. Then I changed the timezone in laravel to + 3 h

then I updated that row and I replicated your behaviour in the hours area (not seconds)

//        dd(Timestamp::find(13876)->update(['time'=>now()]));
        dd(Timestamp::all()->toArray());

![image](https://github.com/laravel/framework/assets/82870454/2f89a6c1-d31c-4277-8b2a-c7e70cb5d298)
